doc/decisions/constructor_functions.md

```diff
@@ -0,0 +1,112 @@
+# Constructor Functions
+
+## Problem
+
+The constructor functions for creating Keys and KeySets must be easy to be invoked from other languages without complicated workarounds.
+In particular, another implementation should not need to resort to adding a small amount of C to "bridge the gap".
```

Comment:
In which binding do we have C code only for constructors? In those we do not have any C code (like JNA), keyNew and ksNew works without any bridge code. In those we already have C code, fixing only the constructors seems pointless to me.

@lawli3t What is the situation in Rust?

Reply:
The Go binding definitely has some wrapper C code already. 

https://github.com/ElektraInitiative/go-elektra/blob/0462a716b69709257ad2e7b0b881eef645a0160f/kdb/key.go#L3-L13

The existing Rust binding doesn't have wrapper code, but it tries it's best to ignore the vararg part of the C API:

https://github.com/ElektraInitiative/libelektra/blob/eac1c6a74cb75bb84ebfae2247b3cc531259ded8/src/bindings/rust/elektra/src/writeable.rs#L55-L75

The Java binding does use the varargs, but needs some pretty horrible translation code:

https://github.com/ElektraInitiative/libelektra/blob/eac1c6a74cb75bb84ebfae2247b3cc531259ded8/src/bindings/jna/libelektra/src/main/java/org/libelektra/Key.java#L116-L137

which at the end just calls another layer of translation code: 

https://github.com/ElektraInitiative/libelektra/blob/eac1c6a74cb75bb84ebfae2247b3cc531259ded8/src/bindings/jna/libelektra/src/main/java/org/libelektra/Key.java#L92-L102

The SWIG bindings are based on C++ and AFAICT the SWIG bindings and C++ are the only ones where the varargs API from C can actually be used.

In summary, maybe the phrasing about "C to bridge the gap" was not quite accurate. It applies in general IMO that there shouldn't be a need for bridge code, but it is not the current problem. Nonetheless, I'd say creating a varargs binding in the core API which then can't always be used directly is a bad idea. That is why I would avoid varargs in the core API and either completely avoid them or at the very least put them into a separate lowlevel C library. It simply allows more flexibility and allows the bindings to better fit their languages.

AFAIK the main reason for the whole varargs API is that it allows creating a whole `KeySet` in a single expression. In C this can only be achieved, if a `KeySet` is created by single function that receives all the `Key`s _and_ a `Key` can also be constructed with a single function. But in other languages there are different, more idomatic possibilities. In Java, Rust and many other languages, a builder pattern makes more sense. I don't see the point in burying an impractical varargs API underneath those builders. Especially not if the varargs API still just calls `keySetName`, `keySetMeta`, etc.

> fixing only the constructors seems pointless to me

To me it seems pointless to have varargs functions. I don't actually know of a single argument in favour of varargs other than single expression thing (which I've now shown multiple times can be achieved differently too). IMO varargs just remove more of the already very lax type checks from the C compiler.

---

At the end there is one thing I am 100% certain of. There is absolutely no place for varargs APIs in the `libelektra-core` API. We can argue about whether or not we want to use them in `libelektra-lowlevel-c` (which targets only C), but for the core API I don't see any reason to have a varargs API.

Reply:
> At the end there is one thing I am 100% certain of. There is absolutely no place for varargs APIs in the libelektra-core API. We can argue about whether or not we want to use them in libelektra-lowlevel-c (which targets only C), but for the core API I don't see any reason to have a varargs API.

If there is no massive performance benefit (with a proper keyNew allocating everything at once), I agree. If there is a massive performance benefit, the status quo makes sense: bindings wrap their keyNew in a way it is suitable for them . E.g. if they have a constructor that always creates key+value, they would call e.g. `keyNew(key, KEY_VALUE, value, KEY_END)`.

So we need a benchmark (that shows that allocations are not the bottleneck of keyNew) to make a proper decision.

Btw. I am very thankful that you are so persisting in these things as this is the only way to look deeply enough to make really good decisions.